network/network.go

```diff
@@ -874,23 +881,47 @@ func (n *network) gossipContainer(chainID, containerID ids.ID, container []byte)
 
 // assumes the stateLock is held.
 func (n *network) track(ip utils.IPDesc) {
-	if n.closed.GetValue() {
+	str := ip.String()
+	if !n.shouldTrack(str) {
 		return
 	}
 
+	// Add [str] to disconnectedIPs to signal to the new goroutine
+	// not to terminate.
+	n.disconnectedIPs[str] = struct{}{}
+	go n.connectTo(ip)
+}
+
+// assumes the stateLock is held
+// return true if [ip] should be tracked
+func (n *network) shouldTrack(str string) bool {
+	// If [str] is already being tracked, do not attempt to track it
+	_, isDisconnected := n.disconnectedIPs[str]
```

Comment:
nit: we can return here if `isDisconnected == true` (similar for `isConnected`, `isMyself`, `closed`)

It'll make the code a little less clean but it'd be a slight performance boost (not that this code is a bottleneck, but can't hurt)